### PR TITLE
[msbuild] Add ILStrip'ing for net6 applications. Fixes #11445.

### LINKED ISSUE #11445: [dotnet] Missing `mono-cil-strip` (or alternative)
Release builds should strip all IL from assemblies (the one that were AOT'ed) to minimize the final app bundle size.

This is done using `mono-cil-strip` for the legacy SDK. However this tools ships with the Mono SDK and there is not, so far, a package (or an alternative tool) to achieve the same for dotnet SDK.

### PR DESCRIPTION
**Pull request by chamons**
- Controlled by EnableAssemblyILStripping which defaults to true
- Integration test included

Before - https://gist.github.com/chamons/c7886f7bacbc2e5ac5966e4251d13e71
After - https://gist.github.com/chamons/148e1bef22fa336f953f3d02dcf20667

859,136 -> 527,872 managed

Fixes https://github.com/xamarin/xamarin-macios/issues/11445.

### CODE REVIEW

**Comment by chamons** [comment]
This is marked as do-not-merge now as the XVS subclass support I just landed part of breaks the build due to a duplicate cecil issue that I'll be mailing about shortly.

**Comment by chamons** [comment]
This is going to require upstream changes from net6 before landing.

File: dotnet/targets/Xamarin.Shared.Sdk.targets
Comment (rolfbjarne) [reply dotnet/targets/Xamarin.Shared.Sdk.targets:612]: Maybe just add a comment? That way we don't remove the `_ComputeStripAssemblyIL` entry in `CreateAppBundleDependsOn` thinking it's redundant.

File: tests/dotnet/UnitTests/TestBaseClass.cs
Comment (rolfbjarne) [inline tests/dotnet/UnitTests/TestBaseClass.cs:123]: I think this only finds files in the root directory, which is not necessarily where assemblies will be for fat apps (nor for macOS or Mac Catalyst, but those shouldn't be stripped anyway). IMHO it would be better to get all assemblies in all subdirectories.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: msbuild/Xamarin.Shared/Xamarin.Shared.targets
Comment (chamons) [reply msbuild/Xamarin.Shared/Xamarin.Shared.targets:1160]: Thoughts on ^?

File: dotnet/targets/Xamarin.Shared.Sdk.targets
Comment (chamons) [reply dotnet/targets/Xamarin.Shared.Sdk.targets:628]: Done

File: tests/dotnet/UnitTests/TestBaseClass.cs
Comment (chamons) [reply tests/dotnet/UnitTests/TestBaseClass.cs]: Done

**Comment by chamons** [comment]
Alright, latest RC2 drop has the fix after Steve showed me how to plug it in  - 🎩 

The main test passed locally, waiting to see how the PR tests go.

Rekicked everyone for review given recent changes.

File: dotnet/targets/Xamarin.Shared.Sdk.targets
Comment (chamons) [reply dotnet/targets/Xamarin.Shared.Sdk.targets]: I think this is fixed.

**Review by dalexsoto — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

File: dotnet/targets/Xamarin.Shared.Sdk.targets
Comment (mauroa) [inline dotnet/targets/Xamarin.Shared.Sdk.targets:640]: Shouldn't we add the Output property here like: 
`<Output TaskParameter="StrippedAssemblies" PropertyName="_StrippedAssemblies" />` and remove the explicit `_StrippedAssemblies` Item declaration above? (line 637)

**Comment by chamons** [comment]
Looks like we are hitting this:

```
make[2]: *** No rule to make target `.stamp-configure-projects-mac'.  Stop.
```

https://github.com/xamarin/xamarin-macios/pull/12776

and

```
[] Try 1/3, non-retryable exception caught. Throwing. Details:
No LastRequestResponse on exception ArtifactExistsException: Artifact all-binlogs already exists for build 5227306. Microsoft.TeamFoundation.Build.WebApi.ArtifactExistsException: Artifact all-binlogs already exists for build 5227306.
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.HandleResponseAsync(HttpResponseMessage response, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)
   at Microsoft.TeamFoundation.Build.WebApi.BuildHttpClientCompatBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken, Func`3 processResponse)
   at Microsoft.TeamFoundation.Build.WebApi.BuildHttpClientCompatBase.SendAsync[T](HttpMethod method, IEnumerable`1 additionalHeaders, Guid locationId, Object routeValues, ApiResourceVersion version, HttpContent content, IEnumerable`1 queryParameters, Object userState, CancellationToken cancellationToken, Func`3 processResponse)
   at Agent.Plugins.BuildServer.AssociateArtifactAsync(Guid projectId, Int32 pipelineId, String name, String jobId, String type, String data, Dictionary`2 propertiesDictionary, CancellationToken cancellationToken)
   at Agent.Plugins.PipelineArtifactServer.<>c__DisplayClass2_1.<<UploadAsync>b__5>d.MoveNext()
```

@rolfbjarne @mandel-macaque - Thoughts? Should I rebuild? Land now? Wait for #12776 to go first?

**Review by mauroa — COMMENTED** [review]
Looks good, we should get the workload from the PR build and test it in VS

**Comment by chamons** [comment]
Oh - i forgot this PR needs XVS blessing before landing.

**Comment by chamons** [comment]
@emaf  / @mauroa - We are ready to get sign off from XVS that this works/doesn't break your build.

File: dotnet/targets/Xamarin.Shared.Sdk.targets
Comment (chamons) [reply dotnet/targets/Xamarin.Shared.Sdk.targets]: Done

**Review by mauroa — APPROVED** [review]

**Comment by mauroa** [comment]
@chamons I'll try it today

**Comment by mauroa** [comment]
@chamons we would first need the below PR lo land in main and then rebase this current PR on top of main so we get that fix:

**Comment by chamons** [comment]
Understood. Blocked on https://github.com/xamarin/xamarin-macios/pull/12804 landing for testing

**Comment by rolfbjarne** [comment]
@mauroa #12804 has been merged into main, and main has been merged into this PR, so this should be good to go for testing now.

**Review by dalexsoto — APPROVED** [review]

**Comment by chamons** [comment]
I had to merge main due to a conflict, which may have broken a changed test. Building locally in parallel with CI.

**Review by mauroa — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Comment by mauroa** [comment]
@dalexsoto @chamons @rolfbjarne I don't see new packages for the latest changes/fixes in the CI build. Did something go wrong? I would need to have them in order to test VS

**Comment by rolfbjarne** [comment]
@mauroa our CI isn't in a nice mood 😒  I started another attempt, hopefully it works this time (third time's the charm!)

File: msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Tasks.Core.csproj
Comment (lambdageek) [inline msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS.Tasks.Core.csproj:30]: FYI, for net7 I'm looking at consolidating the MonoTargets tasks into a single assembly https://github.com/dotnet/runtime/pull/59720

**Comment by chamons** [comment]
1 test failure is known:

image

https://github.com/xamarin/maccore/issues/2443

The other two seem obviously related.

**Comment by chamons** [comment]
@mauroa - I just had to re-merge due to a test file conflict but hopefully the last build is good enough to test today?

**Comment by mauroa** [comment]
@chamons @rolfbjarne @dalexsoto I tested with the latest build and I was able to successfully build a net6 ios app with ILStrip enabled. I think we are good to signoff and merge the PR if everything else is ok

**Comment by rolfbjarne** [comment]
Test failures are unrelated

* monotouch-test/Mac Catalyst [dotnet]/Debug [dotnet]: https://github.com/xamarin/maccore/issues/2443
* dont link/Mac Catalyst [dotnet]/Debug [dotnet]: https://github.com/xamarin/maccore/issues/2414

**Pull request merged**
